automation/clients/airtable.py

```diff
@@ -0,0 +1,225 @@
+"""Tools for interacting with airtable
+
+Includes:
+- Base models, for creating classes from airtable records
+- "Meta" base models
+    + Track the record's status field
+    + Additional JSON metadata for the automation
+- Table specs for associating models, table names, and misc
+- A table poller for executing callbacks on status changes
+
+"""
+
+import abc
+from automation.config import AirtableConfig
+import datetime
+import logging
+from typing import Callable, Dict, Optional, Set, Type
+
+import pydantic
+
+from airtable import Airtable
+
+logger = logging.getLogger(__name__)
+
+
+##########
+# CONSTS #
+##########
+
+DEFAULT_POLL_TABLE_MAX_NUM_RETRIES = 3
+
+
+###############
+# BASE MODELS #
+###############
+
+
+# TODO : consider not allowing users change the id field
+class BaseModel(pydantic.BaseModel, abc.ABC):
+    id: str
+    created_at: datetime.datetime
+
+    class Config:
+        allow_population_by_field_name = True
+
+    @classmethod
+    def from_airtable(cls, raw_dict):
+        return cls(
+            id=raw_dict["id"],
+            created_at=raw_dict["createdTime"],
+            **raw_dict["fields"],
+        )
+
+    def to_airtable(self):
+        fields = self.dict(by_alias=True, exclude_none=True)
+        del fields["id"]
+        del fields["created_at"]
+
+        return {
+            "id": self.id,
+            "fields": fields,
+        }
+
+
+# TODO : should this class inherit from ABC?
+class MetaBaseModel(BaseModel, abc.ABC):
+    meta: Optional[pydantic.Json] = pydantic.Field(default=None, alias="_meta")
+    meta_last_seen_status: Optional[str] = pydantic.Field(
+        default=None, alias="_meta_last_seen_status"
+    )
+    status: Optional[str] = pydantic.Field(default=None, alias="Status")
+
+    @staticmethod
+    @abc.abstractmethod
+    def get_valid_statuses() -> Set[str]:
+        ...
+
+    @pydantic.validator("status", allow_reuse=True)
+    @pydantic.validator("meta_last_seen_status", allow_reuse=True)
+    def validate_status(cls, v):
+        valid_statuses = cls.get_valid_statuses()
+
+        if v not in valid_statuses:
+            raise ValueError(
+                "Status '{}' not in valid statuses: {}".format(
+                    v, ", ".join(valid_statuses)
+                )
+            )
+
+        return v
+
+
+#########
+# TABLE #
+#########
+
+
+class TableSpec(pydantic.BaseModel):
+    name: str
+    model_cls: Type[BaseModel]
+    status_to_cb: Dict[Optional[str], Callable[[BaseModel], BaseModel]]
+
```

Comment:
Unfortunately I changed this file too much for git to recognize it as a rename. Everything above this point is pretty much untouched, everything below is different, in the following ways:

* `TableSpec` lets you get an `AirtableClient` for it. 
* That client then knows about the model class it's dealing with, and the callbacks, so we don't need to pass that in to all its methods anymore.
* Removed some unused methods, made sure all of the fetching methods are generators
* Moved `poll_table` into that class. It now expects to be given a `Config` object it gives the callbacks access to, for now.